### ISSUE TITLE
Common Next.js issues that could use integrated ESLint rules

### ISSUE THREAD

**Issue by leerob** [vercel/next.js#27051]
### Describe the feature you'd like to request

These are common issues developers run into with Next.js. Some are mentioned in the docs, but ideally you don't have to go check the docs.

### Describe the solution you'd like

Instead, ESLint can provide compile-time feedback and suggestions.

- Trying to call an API route from inside `getStaticProps` / `getServerSideProps` leading to [next build failing](https://stackoverflow.com/questions/68153761/nextjs-api-app-works-in-deploy-and-development-mode-but-it-doesnt-when-i-run).
    - Also related error → "[Only absolute URLs are supported on the server](https://stackoverflow.com/questions/68031836/only-absolute-urls-are-supported-error-while-trying-to-deploy-a-next-js-app)"
- Trying to use `getStaticProps` inside of `_app.js`, assuming it works like the other pages, but it doesn't.
- Forgetting to await an async function inside an API route, works fine locally but does out of band work which will cause issues if there’s multiple requests.
    - Code Example

        ```jsx
        // Found on stackoverflow
        export default function handler(req, res) {
            const { method } = req;
            
            switch (method) {
                case 'GET':
        	   // but this is making a fetch and has no await....
                    getRestaurants();
                    break;
                default:
                    res.status(405).end(`Method ${req.method} Not Allowed`);
            }

            function getRestaurants() {
                const restaurants = data;
                res.status(200).json(restaurants);
            }
        }
        ```

- Using `document` or other browser APIs outside of `useEffect`
- File name casing (e.g. `myFile.js` vs `MyFile.js`). This could *potentially* be solved with linting if import statement diffs from the actual file path? Not sure. 

### Describe alternatives you've considered

N/A

**Comment by stefanprobst** [comment]
maybe also a rule for where [global css imports](https://nextjs.org/docs/basic-features/built-in-css-support#adding-a-global-stylesheet) are allowed?

**Comment by Patil2099** [comment]
I would love to work on this @leerob. Can you please assign this to me?

**Comment by leerob** [comment]
@Patil2099 Which one specifically would you like to work on first?

**Comment by tarokawada** [comment]
@leerob I'll work on `Using document or other browser APIs outside of useEffect`

Just to confirm, are you looking to include these interfaces as well https://developer.mozilla.org/en-US/docs/Web/API#interfaces ?

**Comment by JacobLey** [comment]
> Forgetting to await an async function inside an API route, works fine locally but does out of band work which will cause issues if there’s multiple requests.

I'm not sure how you were planning on detecting API/promises, but if it is typescript based it can probably either defer to this existing rule or extend it somehow if necessary:
https://github.com/typescript-eslint/typescript-eslint/blob/v4.28.3/packages/eslint-plugin/docs/rules/no-floating-promises.md

**Comment by housseindjirdeh** [comment]
Not sure if its a super-common issue, but an integrated ESLint rule to catch https://github.com/vercel/next.js/issues/27134 would be ideal.

I'll submit a PR for this soon!

**Comment by leerob** [comment]
Another one: throw an error if using fs or other node packages on the client side. 

https://twitter.com/leeerob/status/1417818534026022912?s=21

**Comment by yordis** [comment]
> File name casing (e.g. myFile.js vs MyFile.js). This could potentially be solved with linting if import statement diffs from the actual file path? Not sure.

We enforce to be kebab-case file name, no more case sensitive vs not issues. Worth reading a bit more here: https://github.com/straw-hat-team/adr/tree/master/adrs/3122196229

I think the problem begins by adopting things that cause more harm than good in the long run. I don't remember anymore how many times I had to fix CI by telling people to fix the casing situation and reminds them to use `git mv` 

:2cents:

**Comment by leerob** [comment]
@yordis I'm not saying Next.js ESLint should have a preference on how you do casing – that's a bit to stylistic to determine at the framework level. I'm talking about when you import a file `MyFile.js` but the _actual_ file name is cased different, like `myFile.js`.

**Comment by yordis** [comment]
Oh like, check for those errors?! Sorry, I misunderstood

**Comment by kyliau** [comment]
Hello everyone! I'm Keen and I just joined Google's WebSDK (Aurora) team this week 😃 
I'd like to tackle the third item in Lee's issue as a starter project:
> Forgetting to await an async function inside an API route, works fine locally but does out of band work which will cause issues if there’s multiple requests.

Please let me know if you have any concerns, otherwise I'll get started!

**Comment by LetItRock** [comment]
hi all! @leerob I would like to try to implement this rule: 
`Another one: throw an error if using fs or other node packages on the client side.`

Just to clarify, the rule should not allow using `node built-in modules` inside the react components, but will allow inside the other files like ex. utils, API routes and inside the data fetching functions`getStaticProps, getStaticPaths, getServerSideProps` is that correct?

**Comment by stefanprobst** [comment]
what do you think about a rule which suggests adding a `key` prop on `meta` elements? (like [this](https://github.com/stefanprobst/next.js/commit/6abb4fdbc8f74a5e2564b8aeb2a2156c3ce42dc0))

**Comment by leerob** [comment]
@stefanprobst Yeah, I think that makes sense - only if there are duplicate elements right? If there's a single `meta` tag, you don't need a `key`, correct? I wonder if we could make it smart enough to warn only if there are duplicates.

**Comment by stefanprobst** [comment]
[the docs on this](https://nextjs.org/docs/api-reference/next/head) are a bit sparse, but if i understand correctly the point is that it's hard to know if somewhere else in the tree the same `meta` property gets set, so it's best to use a `key` prop to avoid duplicate `meta` tags(??)

for example, if you have:

```jsx
// pages/_app.js
import Head from "next/head";
export default function App({ Component, pageProps }) {
  return (
    <>
      <Head>
        <meta property="og:title" content="My app" />
      </Head>
      <Component {...pageProps} />
    </>
  );
}
```

and

```jsx
// pages/index.js
import Head from "next/head";
export default function HomePage() {
  return (
    <>
      <Head>
        <meta property="og:title" content="Home page" />
      </Head>
      <main>
        <h1>Hello world</h1>
      </main>
    </>
  );
}
```

you will end up with *both* `<meta property="og:title"  />` tags in the `head`.

note: `meta` tags with `name` attribute are deduplicated automatically, so this is mostly about `<meta property="..." />` afaiu

**Comment by nguyen-michael** [comment]
@leerob 

Thoughts on a rule that lints if `next/image` is using `layout=` `fill` or `responsive` but isn't setting explicit `sizes` prop?

Docs state:

> A string mapping media queries to device sizes. Defaults to 100vw. We recommend setting sizes when using layout="responsive" or layout="fill" and your image will not be the same width as the viewport.

The rule could be opted into as a warning and reminder to set proper `sizes` or risk loading images that are too large if they're not going to be `100vw`. 

I've got a working implementation of this rule in a custom ESLint plugin stored on a private repo but would happily put it up for review if you think it would be handy!

**Comment by RomkeVdMeulen** [comment]
Maybe as an optional one: ensure `getStaticProps` / `getServerSideProps` are placed above the page component in the source file.

I spent a few hours debugging today before I noticed a `getServerSideProps` on line 150+.

**Comment by rubytree33** [comment]
Are any of these still needed? I'm new to open source (I've already read the contributor guide, don't worry) and thinking about implementing one of these. It seems like no progress has been made in here or the linked PRs in a while.

**Comment by jasongerbes** [comment]
@leerob another suggestion: an ESLint rule that ensure safe usage of environment variables. See #46042.
